Added support for always-auth and _auth. Both for root and scopes #16528

jozemlakar opened the pull request:
I had added the loading of auth config when always-auth and _auth are set.
I had added the sending of authorization header in the mentioned case. 
I have tested this against artifactory, when set globally. Someone else might want to test it against other registries or when scopes are used.

Commit:
Added always-auth and _auth to pacote config. Both for root and scopes
Added auth header when always_auth and _auth are set

snirad commented on lib/config/pacote.js:
```diff
@@ -46,9 +46,17 @@ function pacoteOpts (moreOpts) {
   npm.config.keys.forEach(function (k) {
     const authMatch = k[0] === '/' && k.match(
-      /(.*):(_authToken|username|_password|password|email|always-auth)$/
+      /(.*):(_authToken|username|_password|password|email|always-auth|_auth)$/
```
this regex assume we use 
//registry.npmjs... type of syantax.

could you please change it like so ?

/((.*):)?(_authToken|username|_password|password|email|always-auth|_auth)$/

Commit:
Fixed RegEx to general form

snirad reviewed (commented):
- sorry for the confusion , now i see that it works even without adding this regex, but maybe it is still needed, im not sure.

Commit:
Reverted the RegEx as it was a blind shot.

Commit:
Bad revert

zkat commented on node_modules/pacote/lib/fetchers/registry/fetch.js:
```diff
@@ -94,6 +95,8 @@ function getHeaders (uri, registry, opts) {
     headers.authorization = `Basic ${encoded}`
+  } else if (shouldAuth && auth._auth) {
+    headers.authorization = `Basic ${auth._auth}`
   }
```
These changes should be files as a separate PR against `pacote`.

jozemlakar replied:
Added https://github.com/zkat/pacote/pull/96
This has to be verified and then tagged so we can upgrade dependency in npm. Then I will try and fix the config.

zkat commented on lib/config/pacote.js:
```diff
@@ -46,9 +46,17 @@ function pacoteOpts (moreOpts) {
     Object.assign(opts, ownerStats)
   }
 
+  if (!!npm.config.get('always-auth')) {
```
I'd rather not coerce to boolean here. Let `if` do its usual work?

zkat commented on lib/config/pacote.js:
```diff
@@ -46,9 +46,17 @@ function pacoteOpts (moreOpts) {
+      opts.auth._auth=npm.config.get('_auth');
+    }
+  }
+
```
Please make sure the code passes `standard` linting.

zkat commented on lib/config/pacote.js:
```diff
@@ -46,9 +46,17 @@ function pacoteOpts (moreOpts) {
+  if (!!npm.config.get('always-auth')) {
+    if (!opts.auth) { opts.auth = {} }
+    opts.auth.alwaysAuth = true;
+    if (!!npm.config.get('_auth')) {
```
This can't just be for `_auth`. All of the other auth parameters listed below should be made available as toplevel settings.

zkat reviewed (changes requested):
So far so good! A few things to improve first 👍

snirad commented on lib/config/pacote.js:
```diff
@@ -46,9 +46,17 @@ function pacoteOpts (moreOpts) {
   npm.config.keys.forEach(function (k) {
     const authMatch = k[0] === '/' && k.match(
-      /(.*):(_authToken|username|_password|password|email|always-auth)$/
+      /(.*):(_authToken|username|_password|password|email|always-auth|_auth)$/
```
how about adding this line just below, and making authMatch ; let instead of const.
that should hold for all cases, and you can remove that if condition you added just for _auth

`
  npm.config.keys.forEach(function (k) {
    let authMatch = (k[0] === '/' && k.match(/(.*):(_authToken|username|_password|password|email|always-auth|_auth)$/))
   if(!authMatch && /(_authToken|username|_password|password|email|always-auth|_auth)$/.test(k)){
        authMatch =[,registryKey(opts.registry), k]
}
    if (authMatch) {
      const nerfDart = authMatch[1]
      const key = authMatch[2]
      const val = npm.config.get(k)
.........................

`

- also we will need to import registryKey functoin as wel

`function registryKey (registry) {
  const parsed = url.parse(registry)
  const formatted = url.format({
    host: parsed.host,
    pathname: parsed.pathname,
    slashes: parsed.slashes
  })
  return url.resolve(formatted, '.')
}`

what do you think,
@zkat , @jozemlakar

jozemlakar replied:
sounds good. I am currently working on tests on a change in https://github.com/zkat/pacote and will then do config. Whatever the final solution, it should be the same for scoped and global settings, so my first one is no good.

jozemlakar replied:
So, looked at this issue and here is where I stand:
- PR created for pacote. I think that should clear. Once that lands we can update the dependency in npm
- There is only one general test for config/pacote in npm, so the first thing I need to do is create a test for config parsing
- Fix config parsing so we get all options (including _auth) both on global and scoped.

Commit:
Using new implementation of pacote ahead of updating dependecy

Commit:
Auth options are loaded for scoped and global settings equally

Commit:
linted

jozemlakar commented:
I think this is now complete. It adds support for all auth options, uses options in the same manner when scoped and global.

zkat committed:
shrinkwrap: lean on from field for better forward-compat (#17076)

zkat committed:
deps: use relaxed semver range matching for compatibility (#17007)

Fixes: #16925
Fixes: #16931

zkat committed:
deps: ignore npa.resolve error from parsing peerDeps (#17005)

This emulates npm@4's behavior of simply marking the peerDep
as invalid, instead of crashing.

Fixes: #16981

forivall committed:
shrinkwrap: quick fix for possibility that req is sometimes null (#16937)

PR-URL: https://github.com/npm/npm/pull/16937
Credit: @forivall
Reviewed-By: @forivall

zkat committed:
Merge branch 'release-next' into latest

zkat commented:
Awesome work! Yay tests! I'll merge this once the test suite passes. Thanks a bunch 🎉

zkat commented:
@jozemlakar the tests seem to be failing right now :< https://travis-ci.org/npm/npm/builds/240593681?utm_source=github_status&utm_medium=notification

jozemlakar commented:
@zkat seems that registry configured in test is no longer used but rather goes to default public registry. Will look into it.

jozemlakar commented:
@zkat Build was broken with https://github.com/npm/npm/pull/17076
when merged without running tests first https://github.com/npm/npm/pull/16879/commits/c8528297ada8a542385efbefa6ca0d776bbb6342

Without that change the config is
```
{ type: 'remote',
  registry: undefined,
  where: undefined,
  raw: '@scoped/underscore@http://localhost:1337/scoped-underscore/-/scoped-underscore-1.3.1.tgz',
  name: '@scoped/underscore',
  escapedName: '@scoped%2funderscore',
  scope: '@scoped',
  rawSpec: 'http://localhost:1337/scoped-underscore/-/scoped-underscore-1.3.1.tgz',
  saveSpec: 'http://localhost:1337/scoped-underscore/-/scoped-underscore-1.3.1.tgz',
  fetchSpec: 'http://localhost:1337/scoped-underscore/-/scoped-underscore-1.3.1.tgz',
  gitRange: undefined,
  gitCommittish: undefined,
  hosted: undefined }
```

after change the config is
```
{ type: 'version',
  registry: true,
  where: undefined,
  raw: '@scoped/underscore@1.3.1',
  name: '@scoped/underscore',
  escapedName: '@scoped%2funderscore',
  scope: '@scoped',
  rawSpec: '1.3.1',
  saveSpec: null,
  fetchSpec: '1.3.1',
  gitRange: undefined,
  gitCommittish: undefined,
  hosted: undefined }
```

As a result, npm is sending the request to public npm registry instead of the mock.

I have no idea why this was changed or how this passed existing tests. Please have a look at it.

jozemlakar commented:
@zkat if you need me to check something or try something out, let me know. My knowledge of shrinkwrap is next to zero, so I really can't handle the changes you made on my own.

jozemlakar commented:
@zkat in case I was not understood. Commit https://github.com/npm/npm/commit/c8528297ada8a542385efbefa6ca0d776bbb6342 breaks existing tests i.e. shrinkwrap-scoped-auth.js
When you fix that I can fix other tests I created based on that one.

Commit:
remove specific handling of _auth as not needed

jozemlakar commented:
@snirad you are absolutely correct! I remove the lines, thank you.

Commit:
test results and artefacts added to gitignore

Commit:
Merge remote-tracking branch 'upstream/release-next' into latest

Commit:
Fixed tests following changes from #17076

jozemlakar commented:
@zkat this is now ready. Once i pulled all the changes from release-next and synced my tests with the ones you changed tests succeed.

snirad commented:
@zkat  hi kat :) 

any chance we get to see it on 5.0.4 ?

zkat committed:
unignore things

we don't gitignore these

zkat committed:
removing spurious log statement

zkat commented:
This has been merged into `release-next` and is available in `npmc@5.0.4-canary.4`!
